indi-orion-ssg3/orion_ssg3_ccd.cpp

```diff
@@ -0,0 +1,580 @@
+    CoolerSP[1].fill("COOLER_OFF", "Off", ISS_ON);
+    CoolerSP.fill(getDeviceName(), "CCD_COOLER", "Cooler", MAIN_CONTROL_TAB, IP_RW,
+                       ISR_1OFMANY, 0, IPS_OK);
+
+
+    PrimaryCCD.setMinMaxStep("CCD_EXPOSURE", "CCD_EXPOSURE_VALUE", 0.001, 3600, .001, false);
+    PrimaryCCD.setMinMaxStep("CCD_BINNING", "HOR_BIN", 1, 2, 1, false);
+    PrimaryCCD.setMinMaxStep("CCD_BINNING", "VER_BIN", 1, 2, 1, false);
+
+    addAuxControls();
+
+    return true;
+}
+
+/*******************************************************************************
+** INDI is asking us to update the properties because there is a change in
+** CONNECTION status. This fucntion is called whenever the device is
+** connected or disconnected.
+*******************************************************************************/
+bool SSG3CCD::updateProperties()
+{
+    // Call parent update properties first
+    INDI::CCD::updateProperties();
+
+    if (isConnected())
+    {
+        // Let's get parameters now from CCD
+        setupParams();
+
+        SetTimer(getCurrentPollingPeriod());
+        defineProperty(&GainNP);
+        loadConfig(true, GainNP.getName());
+        defineProperty(&OffsetNP);
+        loadConfig(true, OffsetNP.getName());
+        defineProperty(&CoolerSP);
+        defineProperty(&CoolerPowerNP);
+    }
+    else
+    {
+        deleteProperty(GainNP.getName());
+        deleteProperty(OffsetNP.getName());
+        deleteProperty(CoolerSP.getName());
+        deleteProperty(CoolerPowerNP.getName());
+    }
+
+    return true;
+}
+
+/*******************************************************************************
+** Setting up CCD parameters
+*******************************************************************************/
+void SSG3CCD::setupParams()
+{
+    SetCCDParams(orion_ssg3_get_image_width(&ssg3),
+                 orion_ssg3_get_image_height(&ssg3),
+                 orion_ssg3_get_pixel_bit_size(&ssg3),
+                 orion_ssg3_get_pixel_size_x(&ssg3),
+                 orion_ssg3_get_pixel_size_y(&ssg3));
+
+    /* calculate how much memory we need for the primary CCD buffer */
+    PrimaryCCD.setFrameBufferSize(PrimaryCCD.getXRes() * PrimaryCCD.getYRes() * PrimaryCCD.getBPP() / 8);
+}
+
+/*******************************************************************************
+** Set binning
+*******************************************************************************/
+bool SSG3CCD::UpdateCCDBin(int x, int y)
+{
+    int rc;
+
+    rc = orion_ssg3_set_binning(&ssg3, x, y);
+    if (!rc) {
+        PrimaryCCD.setBin(x, y);
+        return true;
+    }
+    return false;
+}
+
+/**
+ * Client is asking to start exposure
+ * @param duration: The requested exposure duration, in seconds
+ * @return: true if exposure started successfully, false otherwise
+ */
+bool SSG3CCD::StartExposure(float duration)
+{
+    int rc;
+
+    ExposureRequest = duration;
+
+    LOG_DEBUG("Starting exposure");
```

Comment:
Just beware that if users take very short successive exposures, this can swell up the log.